### ISSUE TITLE
Associate a keyboard shortcut with the new "Fuzzy Match" toggle button

### ISSUE THREAD

**Issue by blakeNaccarato** [microsoft/vscode#173866]
### Edit

It looks like there was a misunderstanding when this issue was closed as not planned. The perceived request was "assign a default keyboard shortcut to toggle Fuzzy Match", but what I meant was "_just make it possible at all_ to assign a keyboard shortcut to toggle Fuzzy Match. As it stands, clicking the GUI element for toggling Fuzzy appears to be the only way to do it.

I'm not gonna spam a new issue about this minor nitpick again so soon. If someone stumbles upon this issue and wants to request the feature properly next time, be clearer with a title like, "Allow the user to assign a keyboard shortcut to toggle Fuzzy Match in various places", e.g. in the Explorer filter menu, the "Terminal: Recent command" and "Terminal: Recent directory", etc.

---

### Original comment

A "Fuzzy Match" toggle was introduced in https://github.com/microsoft/vscode/pull/164376, and is [documented in January 2023 release notes](https://code.visualstudio.com/updates/v1_75#_tree-find-contiguous-matching). Would it be possible to assign a keyboard shortcut to this toggle as well?

**Comment by joaomoreno** [comment]
We decided not to have a default keyboard shortcut for this command, out of the box. You can assign your own keyboard shortcuts via <kbd>F1</kbd>, `Preferences: Open Keyboard Shortcuts`. You can also roam them across your machines using Settings Sync.

**Comment by blakeNaccarato** [comment]
I don't mean that a default shortcut should be assigned, but rather that there is no associated command in the Keyboard Shortcuts assignment menu for this GUI action. Or if there is such a command, it doesn't come up by searching "fuzzy" in the Keyboard Shortcuts menu.

**Comment by joaomoreno** [comment]
You're absolutely right! Sorry for dismissing the request too early.

**Comment by Nectres** [comment]
I can start working on it 

I am new to the repository, I just read the code and from what I understand

I have to register an action that toggles the find match type between Fuzzy and Contiguous, right?

**Comment by blakeNaccarato** [comment]
Hi @Nectres, thanks for volunteering as tribute! 😅

Looking near `list.toggleFindMode` is probably a good start. I poked around in the code there because I was curious, as well.

Also, there are fuzzy toggle buttons in the menus that follow from the `workbench.action.terminal.RunRecentCommand` and `workbench.action.terminal.goToRecentDirectory`commands. These fuzzy toggles should probably should have commands associated as well. To my untrained eye, it was tougher to spot where the fuzzy toggle logic was hiding in that spot.

I guess there will have to be a decision on what to call these three newly-minted commands, as well. Unless they can be unified into a single command, but I don't know what patterns should be followed in this regard.

But that's about the limit of my expertise, I'm sure a core dev will be able to sketch a clearer picture!

**Comment by Nectres** [comment]
I have found the logic to achieve this, coded it up and tested it, it works.
Added it as a "Toggle Find Match Type" for now.
![image](https://user-images.githubusercontent.com/12008678/220541909-e0100bc9-b697-44d2-9f52-057424e649d5.png)

I think it would it be better to have it as two commands: "Switch to Fuzzy Find Type" and "Switch to Contiguous Find Type".
It would make more sense from the keyboard shortcut point of view as well.

**Comment by frypf** [comment]
> I think it would be best to split it as two commands: "Switch to Fuzzy Find Type" and "Switch to Contiguous Find Type".
It would make more sense from the keyboard shortcut point of view as well.

Having explicit switching available rather than just a toggle would be very useful. On a related note, would you also be able to add explicit switching for `list.toggleFindMode`?

**Comment by Nectres** [comment]
Yeah, it's nearly the same process. It should be submitted as a separated commit request, right?

**Comment by frypf** [comment]
Probably although I'm not sure so I'll defer to @joaomoreno - it's likely I should have created a separate issue, but coincidentally I was googling `list.toggleFindMode` earlier and this issue thread popped up near the top of the results!

**Comment by Nectres** [comment]
![image](https://user-images.githubusercontent.com/12008678/220562696-612aaf79-2a2e-48c2-8f8e-09929f188d3b.png)
@blakeNaccarato  Does this satisy your request?


@joaomoreno
I have added these three commands, assigned a keyboard shortcut to one of them and tested it as well, they work.
I checked and all the other explorer actions are related to views and focusing on them so there was no `explorerActions.ts` in `vs/workbench/browser/actions`. I have created that file and used the "Explorer" category in all the actions registered.
If this structure is fine, can I open a pull request?

**Comment by blakeNaccarato** [comment]
That looks about right, except the name "Toggle find match type" may be a bit vague.

How about the fuzzy toggle from the "Recent Commands" and "Recent Directories" integrated terminal menu? That's triggered by opening an integrated terminal, being at the command-line, then triggering the "recent command" menu using the command palette.

There's a fuzzy toggle there that might live somewhere else in the codebase, unless it's the same logic path for all of them. And if it is the same logic, I wonder if all the toggles share the same "on/off" state?

**Comment by keerthivasansa** [comment]
Yeah, the commands are operational I have coded them. The logic for it I got it directly from the widget displayed in the explorer file search menu. Then I created three different versions wrapped with the Action2 logic and registered them. 
I will just open the pull request.

**Comment by blakeNaccarato** [comment]
Thanks for that, it looks good @keerthivasansa.

**Comment by ashutosh887** [comment]
I'm willing to work on this

**Comment by Jaswir** [comment]
@keerthivasansa @blakeNaccarato Just looking into this issue, looks like it is solved, but changes requested in PR not executed? 
So is this still open, can we finish this?

**Comment by keerthivasansa** [comment]
> @keerthivasansa @blakeNaccarato Just looking into this issue, looks like it is solved, but changes requested in PR not executed? 
> So is this still open, can we finish this?  

Tbh I don't know what's still pending. I think i have implemented all the requested changes. Can you help me understand what's pending?

**Comment by blakeNaccarato** [comment]
Thanks for the implementation @keerthivasansa and @joaomoreno for your understanding in re-opening the issue, I appreciate it!! A heads up if others find this issue, these `list.toggleFind...` commands pair well with the `listSupportsFind` "when expression", allowing you to deconflict the keybind with other similar commands. See the `keybindings.json` snippet below.

Thanks again!

```JSONC
  // Bind toggle filter and fuzzy find in list search (like Explorer view fuzzy search)
  {
    "key": "ctrl+shift+d",
    "command": "list.toggleFindMode",
    "when": "listSupportsFind"
  },
  {
    "key": "ctrl+shift+f",
    "command": "list.toggleFindMatchType",
    "when": "listSupportsFind"
  },
  {
    "key": "ctrl+shift+f",
    "command": "workbench.action.findInFiles",
    "when": "!listSupportsFind"
  },
  {
    "key": "ctrl+shift+f",
    "command": "-workbench.action.findInFiles"
  },
```
